core/src/main/scala/org/apache/spark/broadcast/TorrentBroadcast.scala

```diff
@@ -85,10 +86,21 @@ private[spark] class TorrentBroadcast[T: ClassTag](obj: T, id: Long)
   /** Total number of blocks this broadcast variable contains. */
   private val numBlocks: Int = writeBlocks(obj)
 
+  /** The checksum for all the blocks. */
+  private var checksums: Array[Int] = _
+
   override protected def getValue() = {
     _value
   }
 
+  private def caclChecksum(block: ByteBuffer): Int = {
+    // block is HeapByteBuffer
+    assert(block.hasArray)
```

Comment:
Hm..., it's surprising to me that this doesn't fail any test. It means we may have some unnecessary memory copy that converts DirectByteBuffer (from Netty) to HeapByteBuffer when fetching from remote block manager.

Anyway, I think it's better to handle the non-array case like this:
```
if (block.hasArray) {
   ...
} else {
   byte[] bytes = new byte[block.remaining()]
   block.duplicate.get(bytes)
   adler.update(bytes)
}

```